File: aurora-scala/src/main/scala/net/gree/aurora/scala/domain/tablename/TableName.scala
package net.gree.aurora.scala.domain.tablename

import net.gree.aurora.domain.tablename.TableNameFactory
import net.gree.aurora.domain.tablename.{TableName => JTableName}
import scala.language.implicitConversions

/**
 * [[net.gree.aurora.domain.tablename.TableName]]のScala版。
 */
trait TableName {

  /**
   * 名前
   */
  val name: String

}

/**
 * コンパニオンオブジェクト。
 */
object TableName {

  /**
   * JavaオブジェクトからScalaオブジェクトを生成する。
   *
   * @param underlying [[net.gree.aurora.domain.tablename.TableName]]
   * @return [[net.gree.aurora.scala.domain.tablename.TableName]]
   */
  private[scala] def apply(underlying: JTableName): TableName =
    new TableNameImpl(underlying)

  /**
   * ファクトリメソッド。
   *
   * @param value 識別子の値
   * @return [[net.gree.aurora.scala.domain.tablename.TableName]]
   */
  def apply(value: String): TableName =
    apply(TableNameFactory.create(value))

  /**
   * ScalaオブジェクトからJavaオブジェクトに変換する。
   *
   * @param self [[net.gree.aurora.scala.domain.tablename.TableName]]
   * @return [[net.gree.aurora.domain.tablename.TableName]]
   */
  implicit def toJava(self: TableName) = self match {
    case s: TableNameImpl => s.underlying
  }

}

File: aurora-core/src/main/java/net/gree/aurora/domain/tablename/TableName.java
package net.gree.aurora.domain.tablename;

/**
 * テーブル名を表すインターフェイス。
 */
public interface TableName {

    /**
     * テーブル名を取得する。
     *
     * @return テーブル名
     */
    String getName();

    /**
     * 文字列がテーブル名と一致するかどうかを検査する。
     *
     * @param value 文字列
     * @return 一致する場合はtrue
     */
    boolean isMatch(String value);

}
